kore/src/Kore/Syntax/Json/Internal.hs

```diff
@@ -0,0 +1,687 @@
+    KJMultiApp{assoc, symbol, sorts, args}
+        | null args -> Prelude.Left $ MissingArg "MultiApp"
+        | otherwise ->
+            withAssoc assoc <$> mkF symbol sorts <*> traverse toParsedPattern args
+  where
+    embedVar ::
+        (VariableName -> SomeVariableName VariableName) ->
+        Id ->
+        Sort ->
+        Either JsonError (Variable (SomeVariableName VariableName))
+    embedVar cons n s =
+        Variable <$> mkVarName cons n <*> mkSort s
+
+    mkVarName ::
+        (VariableName -> SomeVariableName VariableName) ->
+        Id ->
+        Either JsonError (SomeVariableName VariableName)
+    mkVarName embed = pure . embed . koreVar
+
+    toSymbol :: Id -> [Sort] -> Either JsonError Kore.SymbolOrAlias
+    toSymbol n sorts = Kore.SymbolOrAlias (koreId n) <$> traverse mkSort sorts
+
+    withAssoc :: LeftRight -> (a -> a -> a) -> [a] -> a
+    withAssoc Left = foldl1'
+    withAssoc Right = foldr1
+
+    mkOr :: Sort -> Either JsonError (ParsedPattern -> ParsedPattern -> ParsedPattern)
+    mkOr s = do
+        sort <- mkSort s
+        pure (\a b -> embedParsedPattern $ OrF $ Kore.Or sort a b)
+
+    mkF :: Id -> [Sort] -> Either JsonError (ParsedPattern -> ParsedPattern -> ParsedPattern)
+    mkF n sorts = do
+        sym <- toSymbol n sorts -- TODO should maybe check that length ss == 2?
+        pure (\a b -> embedParsedPattern $ ApplicationF $ Kore.Application sym [a, b])
+
+koreId :: Id -> Kore.Id
+koreId (Id name) = Kore.Id name Kore.AstLocationNone
+
+koreVar :: Id -> Kore.VariableName
+koreVar (Id name) =
+    -- TODO check well-formed (initial letter, char. set)
+    VariableName (Kore.Id base Kore.AstLocationNone) suffix
+  where
+    baseName = T.dropWhileEnd isDigit name
+    endDigits = T.takeWhileEnd isDigit name
+    (zeros, actualNum) = T.break (/= '0') endDigits
+    (base, suffix)
+        | T.null endDigits = (baseName, Nothing)
+        | T.null actualNum = (baseName <> T.init zeros, Just $ Element 0)
+        | otherwise =
+            (baseName <> zeros, Just $ Element (read $ T.unpack actualNum))
+
+mkSort :: Sort -> Either JsonError Kore.Sort
+mkSort Sort{name, args} =
+    fmap (Kore.SortActualSort . Kore.SortActual (koreId name)) $
+        mapM mkSort args
+mkSort (SortVariable name) =
+    pure . Kore.SortVariableSort $ Kore.SortVariable (koreId name)
+
+------------------------------------------------------------
+-- writing helper
+
+fromPattern :: Kore.Pattern VariableName ann -> KorePattern
+fromPattern pat =
+    -- forget the annotation and recurse over the term-like PatternF
+    let _ :< patF = Recursive.project pat
+     in fromPatternF patF
+
+fromPatternF :: Kore.PatternF VariableName (Kore.Pattern VariableName ann) -> KorePattern
+fromPatternF = \case
+    AndF Kore.And{andSort, andFirst, andSecond} ->
+        KJAnd
+            { sort = fromSort andSort
+            , first = fromPattern andFirst
+            , second = fromPattern andSecond
+            }
```

Comment:
There is a trick you can do with instances of [Recursive](https://hackage.haskell.org/package/recursion-schemes-5.2.2.2/docs/Data-Functor-Foldable.html#t:Recursive), namely you can use the [cata](https://hackage.haskell.org/package/recursion-schemes-5.2.2.2/docs/Data-Functor-Foldable.html#v:cata) function, which for `Kore.Pattern` should be:

```haskell
cata :: (CofreeF (PatternF var) ann x -> x) -> Pattern var ann -> x
```
That means, we just need to write a function 

```haskell
fromPatternF :: CofreeF (Kore.PatternF VariableName) ann KorePattern -> KorePattern
fromPatternF (_ :< patt) = case patt of
  AndF Kore.And{andSort, first, second} ->
        KJAnd
            { sort = fromSort andSort
            , first
            , second
            }
  ...
  ```
  and `fromPattern` will simply be `cata fromPatternF`
  
This way you remove the mutual recursion between  `fromPattern` and `fromPatternF` and let `cata` do the recursion for you

Reply:
That looks very elegant! Would love to use it (although it will make me forget what actually happens with the `CoFree` here...)
I just tried it, but got stuck on one problem: the `DomainValueF` in that type is more liberal than the textual Kore, so I extracted a string literal on the fly (directly inlined the recursion into the `domainValueChild` with the expectation of a string). I don't see how I would do this with the `cata`.
https://github.com/runtimeverification/haskell-backend/pull/3131/files/74e1d8d72d359291edddc50821fd4faf8f1c504f#diff-ffa0765d097e720acdb8048005535bff42394df5ca200e46d72ffa44e0d6d73fR564-R567

Reply:
think you should be able to do something like

```haskell
DomainValueF (Kore.DomainValue domainValueSort (KJString value))
    KJDv
        { sort = fromSort domainValueSort
        , value
        }

DomainValueF _ -> error "Bad domain value"
```

Reply:
sorry had that wrong. the edit should work

Reply:
oops...nm. the first version was correct after all

```haskell
DomainValueF (Kore.DomainValue domainValueSort (KJString value))
    KJDv
        { sort = fromSort domainValueSort
        , value
        }

DomainValueF _ -> error "Bad domain value"
```

Reply:
...but `KJString` is from the target type `KorePattern`, not from the `Pattern/PatternF` family... :confused: